meilisearch-core/src/query_builder.rs

```diff
@@ -88,97 +138,25 @@ impl<'c, 'f, 'd, 'i> QueryBuilder<'c, 'f, 'd, 'i> {
-                let ands: Vec<_> = ands.iter().map(Cow::deref).collect();
-                Some(sdset::multi::OpBuilder::from_vec(ands).intersection().into_set_buf())
-            }
-            None => None
-        };
 
-        // for each field to retrieve the count for, create an HashMap associating the attribute
-        // value to a set of matching documents. The HashMaps are them collected in another
-        // HashMap, associating each HashMap to it's field.
-        let facet_count_docids = match self.facets {
-            Some(field_ids) => {
-                let mut facet_count_map = HashMap::new();
-                for (field_id, field_name) in field_ids {
-                    let mut key_map = HashMap::new();
-                    for pair in self.index.facets.field_document_ids(reader, field_id)? {
-                        let (facet_key, document_ids) = pair?;
-                        let value = facet_key.value();
-                        key_map.insert(value.to_string(), document_ids);
-                    }
-                    facet_count_map.insert(field_name, key_map);
-                }
-                Some(facet_count_map)
+        // When there is no candidate document in the facet filter, we don't perform the search,
+        // as the result will alway be empty.
+        if let Some(ref facet_filter) = self.facet_filter {
+            if facet_filter.is_empty() {
+                return Ok(SortResult::default())
             }
-            None => None,
-        };
+        }
 
-        match self.distinct {
-            Some((distinct, distinct_size)) => bucket_sort_with_distinct(
-                reader,
-                query,
-                range,
-                facets_docids,
-                facet_count_docids,
-                self.filter,
-                distinct,
-                distinct_size,
-                self.criteria,
-                self.searchable_attrs,
-                self.index.main,
-                self.index.postings_lists,
-                self.index.documents_fields_counts,
-                self.index.synonyms,
-                self.index.prefix_documents_cache,
-                self.index.prefix_postings_lists_cache,
-            ),
-            None => bucket_sort(
-                reader,
-                query,
-                range,
-                facets_docids,
-                facet_count_docids,
-                self.filter,
-                self.criteria,
-                self.searchable_attrs,
-                self.index.main,
-                self.index.postings_lists,
-                self.index.documents_fields_counts,
-                self.index.synonyms,
-                self.index.prefix_documents_cache,
-                self.index.prefix_postings_lists_cache,
-            ),
+        if self.distinct.is_some() {
+            bucket_sort_with_distinct(reader, query, range, self)
+        } else {
+            bucket_sort(reader, query, range, self)
```

Comment:
It seems to me that bucket sort is part query_builder, given its extensive use of self...